Add better error messages for typical SamplerV2 and EstimatorV2 error cases

t-imamichi opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [X] I have added the tests to cover my changes.
- [X] I have updated the documentation accordingly.
- [X] I have read the CONTRIBUTING document.
-->

### Summary

If users want to run a circuit with SamplerV2, they need to call `SamplerV2.run([circuit])` according to the spec.
https://github.com/Qiskit/qiskit/blob/00b0952c47c5b63edff2674b7e5de4b8f5256927/qiskit/primitives/base/base_sampler.py#L166-L167

But, they may miss `[]` but the corresponding error message was not comprehensive as follows. This PR detects such a case and shows more comprehensive message.

```python
from qiskit import QuantumCircuit
from qiskit.circuit.library import RealAmplitudes
from qiskit.primitives import StatevectorSampler

qc = QuantumCircuit(1, 1)
qc.h(0)
qc.measure(0, 0)

sampler = StatevectorSampler()
result = sampler.run([qc]).result()
# OK

result = sampler.run(qc).result()
# NG
```

output
main branch:
> AttributeError: '_SingletonHGate' object has no attribute 'parameters'

this PR:
> ValueError: An invalid Sampler pub-like was given (<class 'qiskit._accelerate.quantum_circuit.CircuitInstruction'>). If you want to run a single circuit, you need to wrap it with `[]` like `sampler.run([circuit])` instead of `sampler.run(circuit)`.

----------------

I also added a similar error message for Estimator V2 as suggested by @ElePT.
```python
from qiskit import QuantumCircuit
from qiskit.circuit.library import RealAmplitudes
from qiskit.primitives import StatevectorEstimator

qc = QuantumCircuit(1, 1)
qc.h(0)
qc.h(0)
qc.h(0)

estimator = StatevectorEstimator()
result = estimator.run([(qc, "Z")]).result()
# OK

result = estimator.run((qc, "Z")).result()
# NG
```

output

main branch
> TypeError: Invalid observable type: <class 'qiskit._accelerate.quantum_circuit.CircuitInstruction'>

this PR
> ValueError: An invalid Estimator pub-like was given (<class 'qiskit.circuit.quantumcircuit.QuantumCircuit'>). If you want to run a single pub-like, you need to wrap it with `[]` like `estimator.run([(circuit, observables, param_values)])` instead of `estimator.run((circuit, observables, param_values))`.

### Details and comments

In addition to the error messages, I fixed random seed for primitives tests (replaced `np.random.rand` with `np.random.default_rng`).

t-imamichi committed:
add an early error for a case

ElePT commented on qiskit/primitives/containers/sampler_pub.py:
```diff
@@ -113,6 +114,13 @@ def coerce(cls, pub: SamplerPubLike, shots: int | None = None) -> SamplerPub:
+            raise ValueError(
+                f"An invalid Sampler pub-like was given ({pub}). If you want to run a circuit, "
+                "you need to wrap a circuit with `[]` like `sampler.run([circuit])` "
+                "instead of `sampler.run(circuit)`."
```
```suggestion
                f"An invalid Sampler pub-like was given ({pub}). If you want to run a single circuit, "
                "you need to wrap the circuit with `[]` like `sampler.run([circuit])` "
                "instead of `sampler.run(circuit)`."
```

ElePT reviewed (commented):
Thanks a lot for this PR, I think it makes a lot of sense. I added a suggestion to make the message even more clear (specifying single circuit). Would it make sense to extend this to the `Estimator` too? I am not sure if there is also a case with a single circuit and observable where this type of issue would happen.

t-imamichi committed:
Update qiskit/primitives/containers/sampler_pub.py

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

t-imamichi committed:
add better message for estimator pub

t-imamichi commented:
I made the following changes since the last review comment.

- Added an error message for EstimtorV2
- Changed to display the object type instead of the object itself because the message can be very long
- Fixed random seed related `np.random.rand`. I noticed a error occurred due to the randomness. https://github.com/Qiskit/qiskit/pull/12031/checks?check_run_id=22772610685

ElePT commented on qiskit/primitives/containers/estimator_pub.py:
```diff
@@ -132,6 +132,15 @@ def coerce(cls, pub: EstimatorPubLike, precision: float | None = None) -> Estima
+        if isinstance(pub, QuantumCircuit):
+            raise ValueError(
+                f"An invalid Estimator pub-like was given ({type(pub)}). "
+                "If you want to run a single pub-like, you need to wrap it with `[]` like "
```
I think I'd prefer reading "pub" than "pub-like" in the second sentence (it also avoids two likes in the sentence).  It's a tiny detail, but I think that the difference between pub and pub-like is a technicality and we can probably use them interchangeably. I expect both concepts to be foreign to the users anyway (pub, pub-like), but the example really clarifies the message, so I am not too set on this comment, either way is fine for me.

```suggestion
                "If you want to run a single pub, you need to wrap it with `[]` like "
```

t-imamichi replied:
Yes, it makes sense to use just `pub` in the sentence.

ElePT reviewed (commented):
Thanks a lot! I agree that exposing the pub could make the message too long, so using the type is a good call. It might still be a bit unexpected for users to see a report of `<class 'qiskit._accelerate.quantum_circuit.CircuitInstruction'>` (for example) as their input, but I think that overall the message is much much clearer and hopefully the suggestion about adding `[ ]` covers a lot of these initial user errors. Mostly LGTM (I left a tiny tiny comment). The changes to the test seeds/rng also look good.

ElePT reviewed (commented):
Sorry, one more question, would a single sampler pub input like `sampler.run((circuit,parameters))` run successfully? or would it also have to be wrapped in `[]` like the estimator one (`sampler.run([(circuit,parameters)])`)? Asking in case this also needs to be covered by the error messages.

t-imamichi commented:
`sampler.run((circuit,parameters))` isn't correct. It raises an error `ValueError: The number of values (0) does not match the number of parameters (X) for the circuit.` (`X` is the actual number of parameters). In this case, the first item is a valid pub-like (`(circuit,parameters)` is an iterable and the first object is `QuantumCircuit`, which is a pub-like), but it is missing parameter values. So, the error message covers the case.
This error message is also raised in the case of `sampler.run((circuit1, circuit2))` where `circuit1` has some parameters.

t-imamichi committed:
Update qiskit/primitives/containers/estimator_pub.py

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

ElePT commented:
> sampler.run((circuit,parameters)) isn't correct. It raises an error ValueError: The number of values (0) does not match the number of parameters (X) for the circuit. (X is the actual number of parameters). In this case, the first item is a valid pub-like ((circuit,parameters) is an iterable and the first object is QuantumCircuit, which is a pub-like), but it is missing parameter values. So, the error message covers the case.
This error message is also raised in the case of sampler.run((circuit1, circuit2)) where circuit1 has some parameters.

I see... it's difficult to tell apart both cases to have a specific message for each case. What do you think if we also add a second part to the error message in `SamplerPub.validate` so that it says the following??:

```
  raise ValueError(
      f"The number of values ({num_parameters}) does not match "
      f"the number of parameters ({self.circuit.num_parameters}) for the circuit. "
      f"Note that if you want to run a single pub, you need to wrap it with `[]` like "
      f"`sampler.run([(circuit, param_values)])` instead of `sampler.run((circuit, param_values))`."
  )
```

t-imamichi committed:
add an additional message

t-imamichi commented:
Thank you for your suggestion. I added your additional message if the number of parameters do not match and `num_parameters == 0`.

t-imamichi commented on qiskit/primitives/containers/estimator_pub.py:
```diff
@@ -217,8 +217,6 @@ def validate(self):
     An Estimator Pub can also be initialized in the following formats which
     will be converted to the full Pub tuple:
 
-    * ``circuit
-    * ``(circuit,)``
```
They are not valid estimator pub-like because estimator pub-like requires observables. They seem to be copies of sampler pub docstrings.

ElePT reviewed (approved):
Thanks a lot for your changes, I think this looks very good now. Approving!
